api/src/test/java/org/openmrs/api/UserServiceTest.java

```diff
@@ -277,6 +281,49 @@ public void createUser_shouldNotAllowSystemIdEqualsUsernameWithLuhnCheckDigit()
 						newUser.getSystemId()));
 		userService.createUser(newUser, SOME_VALID_PASSWORD);
 	}
+	@Test
+	public void createUser_shouldNotAllowCreatingUserWithPrivilegeCurrentUserDoesnotHave() throws IllegalAccessException {
+		User currentUser = new User();
+		Role userRole = new Role("User Adder");
+		userRole.setRole(RoleConstants.AUTHENTICATED);
+		userRole.addPrivilege(new Privilege("Add Users"));
+		currentUser.addRole(userRole);
+		
+		expectedException.expect(APIException.class);
+		expectedException.expectMessage("You must have privilege {0} in order to assign it");
+		
+		withCurrentUserAs(currentUser, () -> {
+			Role role = new Role();
+			role.setRole(RoleConstants.AUTHENTICATED);
+			role.addPrivilege(new Privilege("Custom Privilege"));
+			
+			User u = new User();
+			u.setPerson(new Person());
+
+			u.addName(new PersonName("Benjamin", "A", "Wolfe"));
+			u.setUsername("bwolfe");
+			u.getPerson().setGender("M");
+			
+			u.addRole(role);
+
+			userService.createUser(u, "Openmr5xy");
+		});
+	}
+
+	private void withCurrentUserAs(User user, Runnable internals) throws IllegalAccessException {
+		UserContext userContext = Context.getUserContext();
+		User authenticatedUser = userContext.getAuthenticatedUser();
+		try {
+			FieldUtils.getField(UserContext.class, "user", true).set(userContext, user);
+			internals.run();
+		} finally {
+			FieldUtils.getField(UserContext.class, "user", true).set(userContext, authenticatedUser);
+		}
+	}
+	@Test
+	public void createUser_shouldNotAllowAssigningSuperUserRoleIfCurrentUserDoesnotHaveAssignSystemDeveloperPrivilege() throws IllegalAccessException {
+		
```

Comment:
So the implementation here should:

1. Setup the logged in user as one who does not have the SUPER_USER role and who doesn't have the ASSIGN_SYSTEM_DEVELOPER privilege associated with any role they do have (as in the test I wrote above).

2. In side the `withCurrentUserAs()` function, tries to assign the SUPER_USER role to a new user.

3. We should verify that the message of the generated exception matches the `User.you.must.have.role` message.

Reply:
So am to write tests by the end of today but again may be formating the thrown exception into array. cc @ibacher , @mks-d ,

Reply:
@sherrif10  , your PR is still noisy  as well. I also good practice is to write tests first , fail them then make them pass with your new changes.